### PR TITLE
[WIP] Skywire Android port

### LINKED ISSUE #138: Port Skywire to Android
**Feature description**

Port Skywire lite with VPN functionality to Android. 

Depends on: 

- #100 

Preliminary research into obstacles and ideas: 

#99

### PR DESCRIPTION
**Pull request by Darkren**
Did you run `make format && make check`?
Yes

Fixes #138 	

 Changes:	
- Added a separate VPN client for mobile devices
- Added `skywiremob` library to expose Skywire network API to mobile devices
- Added example Android Skywire VPN application
- Visor's startup is now fully interruptable

How to test this PR:
1. Install gomobile from https://github.com/golang/mobile
2. Install Android SDK and NDK. Put its locations to corresponding env vars: ANDROID_HOME for SDK and ANDROID_NDK_HOME for NDK
3. Compile the `skywiremob` library. To do this one should put the skywire repository according to the `GOPATH`, i.e.: `$GOPATH/src/github.com/skycoin/skywire`. Then you need to `cd` to the directory right outside of the `GOPATH`. So, having default `GOPATH` at `$HOME/go` you need to `cd $HOME`. From there you may build the library like this:
```
$ gomobile bind -o ./go/src/github.com/skycoin/skywire/cmd/skywirevisormobile/android/app/skywire.aar -target=android ./go/src/github.com/skycoin/skywire/pkg/skywiremob/
```
Probably it may be successfully built the other way, but this is the only way I got it to work. The above command will put the `skywire.aar` library right to where the Android project is looking for it.
4. Build the apk. May be done from the terminal or Android studio
5. Install and run apk
6. Put correct remote PK of visor running VPN server and passcode for that server and press `Start`. Once VPN is ready you will see the toast message
7. Check your IP address

### CODE REVIEW

File: cmd/skywirevisormobile/android/README.md [hunk 1]
```diff
@@ -0,0 +1,57 @@
...
+- `TUNGateway() string`: Requires `ShakeHands` to be called first. Returns the assigned TUN gateway;
+- `StopVisor() string`: Stops currently running visor. Returns non-empty string with error in case of failure;
+- `WaitForVisorToStop() string`: Blocks until visor exits. Returns non-empty string with error in case of failure;
+- `SetMobileAppAddr(string)`: Passes address of the UDP connection opened on the mobile application side;
+- `ServeVPN()`: Starts off the goroutine serving VPN connection. After this call `IsVPNReady` starts returning `true`;
+- `StartListeningUDP() string`: Opens UDP listener on the Go side. Returns non-empty string with error in case of failure.
+
+## Mobile/Go Communication
+API may seem a bit complicated at first. Currently tested for Android devices, should be used with caution on iOS. 
+Mobile app communicate with the Go part via UDP. All the packets are sent to the Go part via UDP and then get resent
```
Comment (nkryuchkov) [inline cmd/skywirevisormobile/android/README.md]: Did you mean `Mobile app communicates`?

**Review by nkryuchkov — APPROVED** [review]
I haven't tested it on an Android phone, but this looks great!

**Comment by Senyoret1** [comment]
Very good job, I was able to use the vpn in an Android device using a local visor. I started polishing some parts of the Android client and added in https://github.com/skycoin/skywire/issues/472 ideas for some improvements that would be good to add to skywiremob for the next versions

**Comment by Darkren** [comment]
Hey there @Senyoret1 I think I'll be adding changes right to this PR for now. Since it's not merged anyway. Starting with #472

File: cmd/skywirevisormobile/android/README.md (hunk 1 above)
Comment (Darkren) [reply cmd/skywirevisormobile/android/README.md]: That's right, will fix

**Pull request closed without merging**